File: exercises/python/starting/cards/snap.py
from animal_deck import AnimalDeck


class Snap:

    def __init__(self, deck):
        self.player_1_score = None
        self.player_2_score = None
        self.deck = deck
        deck.shuffle()

    # def main_method(self):
    #     snap = Snap(AnimalDeck())
    #     snap.play()

    def play(self):
        previous_card = None
        while len(self.deck.cards) > 0:
            current_card = self.deck.deal()
            player_input = str(input("Your turn: "))
            if len(player_input) > 0 and player_input[0] == 'a':
                if current_card.snap(previous_card):
                    print("SNAP! score Player 1")
                    player_1_score += 1
                else:
                    print("WRONG! deducting score from Player 1")
                    player_1_score -= 1
            elif len(player_input) > 0 and player_input[0] == 'l':
                if current_card.snap(previous_card):
                    print("SNAP! scorePlayer 2")
                    player_2_score += 1
                else:
                    print("WRONG! deducting score from Player 2")
                    player2Score -= 1

            previous_card = current_card

        if player_1_score == player_2_score:
            print("Draw")
        elif player_1_score > player_2_score:
            print("Player 1 wins!")
        else:
            print("Player 2 wins!")

        print("Scores: " + player_1_score + " vs " + player_2_score)


def play_snap():
    snap = Snap(AnimalDeck())
    snap.play()


play_snap()

File: exercises/php/cards/snap.php
<?php

/*
 * To execute this PHP script, use the command `php snap.php`
 * from the command line positioned at the `cards/` directory.
 * E.g. cd cards/ && php snap.php
 */


use McrDigital\PhpFundamentals1\Cards\CAnimalDeck;

require_once '../vendor/autoload.php';

class CSnap
{
    private int $player1Score;
    private int $player2Score;
    private CAnimalDeck $deck;

    public function __construct(CAnimalDeck $deck)
    {
        $this->player1Score = 0;
        $this->player2Score = 0;
        $this->deck = $deck;
        $this->deck->shuffle();
    }

    public function play(): void
    {
        $previousCard = null;

        while (count($this->deck->getCards()) > 0) {
            $currentCard = $this->deck->deal();

            echo $currentCard . PHP_EOL;

            $prompt = readline("Play snap: ");

            if (strlen($prompt) > 0 && strtolower(substr($prompt, 0)) === 'a') {
                if ($currentCard->snap($previousCard)) {
                    echo "SNAP! score Player 1" . PHP_EOL;
                    $this->player1Score++;
                } else {
                    echo "WRONG! deducting score from Player 1" . PHP_EOL;
                    $this->player1Score--;
                }
            } else if (strlen($prompt) > 0 && strtolower(substr($prompt, 0)) === 'l') {
                if ($currentCard->snap($previousCard)) {
                    echo "SNAP! score Player 2" . PHP_EOL;
                    $this->player2Score++;
                } else {
                    echo "WRONG! deducting score from Player 2" . PHP_EOL;
                    $this->player2Score--;
                }
            }

            $previousCard = $currentCard;
        }

        if ($this->player1Score === $this->player2Score) {
            echo "Draw..." . PHP_EOL;
        } else if ($this->player1Score > $this->player2Score) {
            echo "Player 1 wins!" . PHP_EOL;
        } else {
            echo "Player 2 wins!" . PHP_EOL;
        }

        echo "Scores: " . $this->player1Score . " vs " . $this->player2Score . PHP_EOL;
    }
}

$animalDeck = new CAnimalDeck();
$snap = new CSnap($animalDeck);
$snap->play();